### PR TITLE
Check if method getLocale exists before to use it

### LINKED ISSUE #297: Attempted to call an undefined method named "getLocale" of class "App\Entity\User"
#### Sonata packages

```
sonata-project/admin-bundle              3.51.0 3.51.0 The missing Symfony Admin Generator
sonata-project/block-bundle              3.15.0 3.15.0 Symfony SonataBlockBundle
sonata-project/cache                     2.0.1  2.0.1  Cache library
sonata-project/core-bundle               3.17.0 3.17.0 Symfony SonataCoreBundle
sonata-project/datagrid-bundle           2.5.0  3.0.0  Symfony SonataDatagridBundle
sonata-project/doctrine-extensions       1.3.0  1.3.0  Doctrine2 behavioral extensions
sonata-project/doctrine-orm-admin-bundle 3.9.0  3.10.0 Symfony Sonata / Integrate Doctrine ORM into the SonataAdminBundle
sonata-project/exporter                  2.0.1  2.0.1  Lightweight Exporter library
sonata-project/intl-bundle               2.6.0  2.6.0  Symfony SonataIntlBundle
sonata-project/translation-bundle        2.4.2  2.4.2  SonataTranslationBundle
```

#### Symfony packages

```
symfony/asset                      v4.3.2  v4.3.2  Symfony Asset Component
symfony/browser-kit                v4.3.2  v4.3.2  Symfony BrowserKit Component
symfony/cache                      v4.3.2  v4.3.2  Symfony Cache component with PSR-6, PSR-16, and tags
symfony/cache-contracts            v1.1.5  v1.1.5  Generic abstractions related to caching
symfony/config                     v4.3.2  v4.3.2  Symfony Config Component
symfony/console                    v4.3.2  v4.3.2  Symfony Console Component
symfony/css-selector               v4.3.2  v4.3.2  Symfony CssSelector Component
symfony/debug                      v4.3.2  v4.3.2  Symfony Debug Component
symfony/debug-bundle               v4.3.2  v4.3.2  Symfony DebugBundle
symfony/debug-pack                 v1.0.7  v1.0.7  A debug pack for Symfony projects
symfony/dependency-injection       v4.3.2  v4.3.2  Symfony DependencyInjection Component
symfony/doctrine-bridge            v4.3.2  v4.3.2  Symfony Doctrine Bridge
symfony/dom-crawler                v4.3.2  v4.3.2  Symfony DomCrawler Component
symfony/dotenv                     v4.3.2  v4.3.2  Registers environment variables from a .env file
symfony/event-dispatcher           v4.3.2  v4.3.2  Symfony EventDispatcher Component
symfony/event-dispatcher-contracts v1.1.5  v1.1.5  Generic abstractions related to dispatching event
symfony/expression-language        v4.3.2  v4.3.2  Symfony ExpressionLanguage Component
symfony/filesystem                 v4.3.2  v4.3.2  Symfony Filesystem Component
symfony/finder                     v4.3.2  v4.3.2  Symfony Finder Component
symfony/flex                       v1.4.4  v1.4.5  Composer plugin for Symfony
symfony/form                       v4.3.2  v4.3.2  Symfony Form Component
symfony/framework-bundle           v4.3.2  v4.3.2  Symfony FrameworkBundle
symfony/http-client                v4.3.2  v4.3.2  Symfony HttpClient component
symfony/http-client-contracts      v1.1.5  v1.1.5  Generic abstractions related to HTTP clients
symfony/http-foundation            v4.3.2  v4.3.2  Symfony HttpFoundation Component
symfony/http-kernel                v4.3.2  v4.3.2  Symfony HttpKernel Component
symfony/inflector                  v4.3.2  v4.3.2  Symfony Inflector Component
symfony/intl                       v4.3.2  v4.3.2  A PHP replacement layer for the C intl extension that includes additional data from the ICU library.
symfony/maker-bundle               v1.12.0 v1.12.0 Symfony Maker helps you create empty commands, controllers, form classes, tests and more so you can forget about writing boilerplate code.
symfony/mime                       v4.3.2  v4.3.2  A library to manipulate MIME messages
symfony/monolog-bridge             v4.3.2  v4.3.2  Symfony Monolog Bridge
symfony/monolog-bundle             v3.4.0  v3.4.0  Symfony MonologBundle
symfony/options-resolver           v4.3.2  v4.3.2  Symfony OptionsResolver Component
symfony/orm-pack                   v1.0.6  v1.0.6  A pack for the Doctrine ORM
symfony/phpunit-bridge             v4.3.2  v4.3.2  Symfony PHPUnit Bridge
symfony/polyfill-intl-icu          v1.11.0 v1.11.0 Symfony polyfill for intl's ICU-related data and classes
symfony/polyfill-intl-idn          v1.11.0 v1.11.0 Symfony polyfill for intl's idn_to_ascii and idn_to_utf8 functions
symfony/polyfill-mbstring          v1.11.0 v1.11.0 Symfony polyfill for the Mbstring extension
symfony/polyfill-php72             v1.11.0 v1.11.0 Symfony polyfill backporting some PHP 7.2+ features to lower PHP versions
symfony/polyfill-php73             v1.11.0 v1.11.0 Symfony polyfill backporting some PHP 7.3+ features to lower PHP versions
symfony/process                    v4.3.2  v4.3.2  Symfony Process Component
symfony/profiler-pack              v1.0.4  v1.0.4  A pack for the Symfony web profiler
symfony/property-access            v4.3.2  v4.3.2  Symfony PropertyAccess Component
symfony/property-info              v4.3.2  v4.3.2  Symfony Property Info Component
symfony/routing                    v4.3.2  v4.3.2  Symfony Routing Component
symfony/security-acl               v3.0.2  v3.0.2  Symfony Security Component - ACL (Access Control List)
symfony/security-bundle            v4.3.2  v4.3.2  Symfony SecurityBundle
symfony/security-core              v4.3.2  v4.3.2  Symfony Security Component - Core Library
symfony/security-csrf              v4.3.2  v4.3.2  Symfony Security Component - CSRF Library
symfony/security-guard             v4.3.2  v4.3.2  Symfony Security Component - Guard
symfony/security-http              v4.3.2  v4.3.2  Symfony Security Component - HTTP Integration
symfony/serializer                 v4.3.2  v4.3.2  Symfony Serializer Component
symfony/serializer-pack            v1.0.2  v1.0.2  A pack for the Symfony serializer
symfony/service-contracts          v1.1.5  v1.1.5  Generic abstractions related to writing services
symfony/stopwatch                  v4.3.2  v4.3.2  Symfony Stopwatch Component
symfony/swiftmailer-bundle         v3.2.8  v3.2.8  Symfony SwiftmailerBundle
symfony/templating                 v4.3.2  v4.3.2  Symfony Templating Component
symfony/test-pack                  v1.0.6  v1.0.6  A pack for functional and end-to-end testing within a Symfony app
symfony/translation                v4.3.2  v4.3.2  Symfony Translation Component
symfony/translation-contracts      v1.1.5  v1.1.5  Generic abstractions related to translation
symfony/twig-bridge                v4.3.2  v4.3.2  Symfony Twig Bridge
symfony/twig-bundle                v4.3.2  v4.3.2  Symfony TwigBundle
symfony/validator                  v4.3.2  v4.3.2  Symfony Validator Component
symfony/var-dumper                 v4.3.2  v4.3.2  Symfony mechanism for exploring and dumping PHP variables
symfony/var-exporter               v4.3.2  v4.3.2  A blend of var_export() + serialize() to turn any serializable data structure to plain PHP code
symfony/web-link                   v4.3.2  v4.3.2  Symfony WebLink Component
symfony/web-profiler-bundle        v4.3.2  v4.3.2  Symfony WebProfilerBundle
symfony/web-server-bundle          v4.3.2  v4.3.2  Symfony WebServerBundle
symfony/webpack-encore-bundle      v1.6.2  v1.6.2  Integration with your Symfony app & Webpack Encore!
symfony/yaml                       v4.3.2  v4.3.2  Symfony Yaml Component
```

#### PHP version

```
PHP 7.3.6-1+ubuntu18.04.1+deb.sury.org+1 (cli) (built: May 31 2019 11:06:48) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.3.6, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.3.6-1+ubuntu18.04.1+deb.sury.org+1, Copyright (c) 1999-2018, by Zend Technologies
```

## Subject

I install the translation bundle and intl bundle to use the Gemo bundle.
Now, after I log in, I get this error. Everything else is fine.

The issue come from the UserLocaleListener
```
public function onInteractiveLogin(InteractiveLoginEvent $event)
    {
        $user = $event->getAuthenticationToken()->getUser();
        if (null !== $user->getLocale()) {
            $this->session->set('_locale', $user->getLocale());
        }
    }
```
My $user does not have `getLocale` function, and I don't want to.

I think we should add something like
```
method_exists($user,'getLocale')
```

### PR DESCRIPTION
**Pull request by VincentLanglet**
## Subject
I am targeting this branch, because this is a bug fix.

Closes #297 

## Changelog
```markdown
### Fixed
Check if method getLocale exists before to use it in UserLocaleListener
```

### CODE REVIEW

**Comment by VincentLanglet** [comment]
@phansys I don't understand the failing tests. They have nothing in common with my change...

**Comment by phansys** [comment]
>@phansys I don't understand the failing tests. They have nothing in common with my change...

Yes, the failing test was caused by a connection issue. I've restarted the job and it's passing now.

You should add a test covering this change.

File: tests/EventSubscriber/LocalizedUser.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+use Symfony\Component\Security\Core\User\UserInterface;
+
+/**
+ * @author Jonathan Vautrin <jvautrin@pro-info.be>
+ */
+class LocalizedUser extends User implements UserInterface
+{
+    private $locale;
+
+    public function __construct($locale)
```
Comment (phansys) [inline tests/EventSubscriber/LocalizedUser.php]:
```suggestion
    public function __construct(string $locale)
```

File: tests/EventSubscriber/UserLocaleSubscriberTest.php [hunk 2]
```diff
@@ -28,21 +28,33 @@ class UserLocaleSubscriberTest extends TestCase
 {
     /**
      * Check if session locale is set to user locale at login.
+     *
+     * @dataProvider userLocaleSubscriberDataProvider
      */
-    public function testUserLocaleSubscriber()
+    public function testUserLocaleSubscriber($user, $expectedLocale)
```
Comment (phansys) [inline tests/EventSubscriber/UserLocaleSubscriberTest.php]:
```suggestion
    public function testUserLocaleSubscriber(UserInterface $user, string $expectedLocale): void
```

File: tests/EventSubscriber/UserLocaleSubscriberTest.php [hunk 3]
```diff
@@ -28,21 +28,33 @@ class UserLocaleSubscriberTest extends TestCase
...
         $userLocaleSubscriber->onInteractiveLogin($event);
-        $this->assertSame('fr', $session->get('_locale'));
+        $this->assertSame($expectedLocale, $session->get('_locale'));
     }
 
-    private function getEvent(Request $request, User $user)
+    /**
+     * @return array
+     */
+    public function userLocaleSubscriberDataProvider()
```
Comment (phansys) [inline tests/EventSubscriber/UserLocaleSubscriberTest.php]:
```suggestion
    public function userLocaleSubscriberDataProvider(): array
```

File: tests/EventSubscriber/UserLocaleSubscriberTest.php [hunk 4]
```diff
@@ -28,21 +28,33 @@ class UserLocaleSubscriberTest extends TestCase
...
-        $user = new User('fr');
         $event = $this->getEvent($request, $user);
         $userLocaleSubscriber = new UserLocaleSubscriber($session);
         $userLocaleSubscriber->onInteractiveLogin($event);
-        $this->assertSame('fr', $session->get('_locale'));
+        $this->assertSame($expectedLocale, $session->get('_locale'));
     }
 
-    private function getEvent(Request $request, User $user)
+    /**
```
Comment (phansys) [inline tests/EventSubscriber/UserLocaleSubscriberTest.php]:
```suggestion
```

File: tests/EventSubscriber/UserLocaleSubscriberTest.php [hunk 5]
```diff
@@ -28,21 +28,33 @@ class UserLocaleSubscriberTest extends TestCase
...
+     * @return array
```
Comment (phansys) [inline tests/EventSubscriber/UserLocaleSubscriberTest.php]:
```suggestion
```

File: tests/EventSubscriber/UserLocaleSubscriberTest.php [hunk 6]
```diff
@@ -28,21 +28,33 @@ class UserLocaleSubscriberTest extends TestCase
...
+     */
```
Comment (phansys) [inline tests/EventSubscriber/UserLocaleSubscriberTest.php]:
```suggestion
```

File: tests/EventSubscriber/UserLocaleSubscriberTest.php [hunk 7]
```diff
@@ -28,21 +28,33 @@ class UserLocaleSubscriberTest extends TestCase
...
+    public function userLocaleSubscriberDataProvider()
+    {
+        return [
+            [new LocalizedUser('fr'), 'fr'],
+            [new User(), 'en'],
+        ];
+    }
+
+    private function getEvent(Request $request, $user)
```
Comment (phansys) [inline tests/EventSubscriber/UserLocaleSubscriberTest.php]:
```suggestion
    private function getEvent(Request $request, UserInterface $user): Event
```

**Comment by VincentLanglet** [comment]
Should be now ok.

File: tests/EventSubscriber/UserLocaleSubscriberTest.php [hunk 8]
```diff
@@ -28,21 +29,30 @@ class UserLocaleSubscriberTest extends TestCase
...
-    private function getEvent(Request $request, User $user)
+    public function userLocaleSubscriberDataProvider(): array
+    {
+        return [
+            [new LocalizedUser('fr'), 'fr'],
+            [new User(), 'en'],
+        ];
+    }
+
+    private function getEvent(Request $request, UserInterface $user)
```
Comment (phansys) [inline tests/EventSubscriber/UserLocaleSubscriberTest.php]: Could you please declare the return type?

Comment (VincentLanglet) [reply tests/EventSubscriber/UserLocaleSubscriberTest.php]: Done

**Comment by phansys** [comment]
Great! :+1:
Could you please squash your commits?

**Comment by OskarStark** [comment]
Thank you, I squashed your commits into one 👍

**Comment by VincentLanglet** [comment]
Thanks @OskarStark

**Pull request merged**